stm/stm.ml

```diff
@@ -23,6 +23,53 @@ open Feedback
 open Vernacexpr
 open Vernac_classifier
 
+exception End_of_input
+module Parser : sig
+  type state
+
+  val init : unit -> state
+  val cur_state : unit -> state
+  val install : state -> unit
+
+  val set_proof_mode : string -> state -> state
+  val unset_proof_mode : state -> state
+
+  val parse :
+    state -> Pcoq.Gram.coq_parsable -> Vernacexpr.vernac_expr Loc.located
+
+  val print : state -> string
+
+end = struct
+
+  open Pcoq
+
+  type state = Pcoq.state
+
+  let init = Pcoq.freeze
+
+  let cur_state () = freeze ()
+  let install x = unfreeze x 
+
+  let on_ps ps f x = unfreeze ps; f x; freeze ()
+  let set_proof_mode mode (ps : state) = on_ps ps set_tactic_entry mode
+  let unset_proof_mode (ps : state) = on_ps ps unset_tactic_entry ()
+
+  let parse ps pa =
+    unfreeze ps;
+    Flags.with_option Flags.we_are_parsing (fun () ->
```

Comment:
My understanding is that Pcoq tries to understand which actions need to be done/undone already in order to change state.
Maybe such check could be made faster, but I would put in Pcoq, not here.
Is it relevant performance wise? I've no clue, maybe we could start a pundulum job?